### ISSUE TITLE
Handle client timeout

### ISSUE THREAD

**Issue by georgefst** [georgefst/monpad#22]
We want to be able to kill threads if they don't respond to pings, and run `onDroppedConnection`. Without this, the Linux backend can end up with an unbounded number of left over devices created in `/dev/input`.

`websockets` recently added [serverRequirePong](https://hackage.haskell.org/package/websockets-0.12.7.2/docs/Network-WebSockets.html#t:ServerOptions) (https://github.com/jaspervdj/websockets/pull/199), which handles some of this logic, though it doesn't allow us to specify an action to run on timeout (EDIT: that might be fine - assuming it just causes an exception to be thrown, we'll end up running `onDroppedConnection` anyway).

Another issue is that `servant-websockets` doesn't allow us to get at `ServerOptions` ~or `ConnectionOptions` (EDIT: not true, we actually now do in 2d3ea6b)~.`

We could manually implement a workaround like the one discussed [here](https://github.com/jaspervdj/websockets/issues/159), though it's pretty ugly.

**Comment by georgefst** [comment]
> the Linux backend can end up with an unbounded number of left over devices

Actually, we shouldn't need to be able to attach an action here to solve this. We just need to kill the thread. Since the devices get closed on garbage collection.

**Comment by georgefst** [comment]
> Another issue is that servant-websockets doesn't allow us to get at `ServerOptions`.

This shouldn't be a surprise. With Servant, we're running a WAI server, rather than using the simple `runServer` provided by the `websockets` library.

**Comment by georgefst** [comment]
In fact, based on discussions in https://github.com/yesodweb/wai/issues/310 and https://github.com/yesodweb/wai/issues/232, it would appear that WAI already sets all this up.

It's possible I only assumed all this was an issue because of https://github.com/composewell/streamly/issues/1203.

**Comment by georgefst** [comment]
Yep, I think this is closed by d9929bdbcc0f3c7578314b7cc115a8008e0a31f9. We actually get visible errors now when e.g. the browser tab is closed.

But I'll keep this open until I've actually confirmed that timeout errors occur.

**Comment by georgefst** [comment]
> But I'll keep this open until I've actually confirmed that timeout errors occur.

Okay, from what I've seen so far, I'm pretty sure they *don't*.

**Comment by georgefst** [comment]
> We just need to kill the thread. Since the devices get closed on garbage collection.

I see no sign of this. There are no guarantees of promptness, but maybe the thread isn't killed?

**Comment by georgefst** [comment]
We should think about how we want to handle reconnections (especially quick ones, e.g. in the case of WiFi dropping). I guess we don't want to immediately kill the thread anyway.

**Comment by domenkozar** [comment]
See also https://github.com/digitallyinduced/ihp/blob/dbb4ec64fe7b460fd80041e0b7a2867d90529d78/IHP/WebSocket.hs#L124
